File: packages/swift-parser/Sources/SwiftParser/DaemonProtocol.swift
import Foundation

enum DaemonProtocol {
    /// Read a single length-prefixed frame from stdin.
    /// Returns nil on EOF.
    static func readFrame(from input: FileHandle) -> Data? {
        let lenData = input.readData(ofLength: 4)
        guard lenData.count == 4 else { return nil }
        let len = lenData.withUnsafeBytes { ptr in
            Int(ptr.load(as: UInt32.self).bigEndian)
        }
        guard len >= 0, len <= 100_000_000 else {
            fputs("Invalid frame length: \(len)\n", stderr)
            return nil
        }
        var payload = Data()
        while payload.count < len {
            let chunk = input.readData(ofLength: len - payload.count)
            if chunk.isEmpty { break }
            payload.append(chunk)
        }
        guard payload.count == len else {
            fputs("Truncated frame: expected \(len) bytes, got \(payload.count)\n", stderr)
            return nil
        }
        return payload
    }

    /// Write a length-prefixed frame to stdout and flush.
    static func writeFrame(to output: FileHandle, payload: Data) {
        var len = UInt32(payload.count).bigEndian
        let lenData = Data(bytes: &len, count: 4)
        output.write(lenData)
        output.write(payload)
    }
}

File: packages/objc-parser/src/daemon_protocol.c
#include "daemon_protocol.h"
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>  // for ntohl/htonl

char* read_frame(FILE* input, uint32_t* out_len) {
    uint32_t net_len;
    if (fread(&net_len, 4, 1, input) != 1) {
        return NULL;  // EOF
    }
    uint32_t len = ntohl(net_len);
    if (len > 100000000) {
        fprintf(stderr, "Invalid frame length: %u\n", len);
        return NULL;
    }
    char* buf = malloc(len + 1);
    if (!buf) return NULL;
    if (fread(buf, 1, len, input) != len) {
        free(buf);
        return NULL;
    }
    buf[len] = '\0';
    *out_len = len;
    return buf;
}

void write_frame(FILE* output, const char* payload, uint32_t len) {
    uint32_t net_len = htonl(len);
    fwrite(&net_len, 4, 1, output);
    fwrite(payload, 1, len, output);
    fflush(output);
}
